Frontend: friends list with React Query

darrenvong opened the pull request:
Lays out the structure for #291 with the initial friend list code implemented with React Query.

This feature is surprisingly complex enough to allow me to play around React Query and demo a few things:
- Dependent queries (i.e. we only fetch the required users after we find out who the user's friend (ID)s are)
- Dynamic number of queries with the new `useQueries` (plural) hook. Dynamic as it depends how many friends a user has!
- Persisting the cache in local storage with the experimental `react-query/persist-localstorage-experimental` package built into React Query
- Devtools from React Query in dev mode that lets you see what's going on - i.e. when React Query refetches and marks a query as stale etc.

There are lots of options to tweak if certain defaults are too "aggressive", but I find them to be working ok so far in this demo, that means it's zero config as promised in the docs so far!

darrenvong committed:
feat: Show list of current user's friends

darrenvong committed:
feat: configure Storybook to use app theme

Closes #403

darrenvong committed:
feat: handle errors/loading state in friend list

darrenvong committed:
feat: share common friends card styles

darrenvong committed:
feat: nest "friends" under "connections" feature + tab in UI

darrenvong committed:
feat: set up React Query

Initial client at app route as recommended + set up dev tools
that come with React Query!

darrenvong committed:
feat: implement friend list using React Query

lucaslcode commented on app/frontend/src/features/connections/friends/FriendList.tsx:
```diff
@@ -46,52 +43,33 @@ const useStyles = makeStyles((theme) => ({
-          friendIds.map((friendId) => {
-            const friend = users[friendId]?.user;
+          friendQueries.map((payload) => {
+            const user = payload.data as User.AsObject;
```
Can we avoid `as`?

darrenvong replied:
Sadly not in this case, since `useQueries` is a generic function that accepts any type of queries, React Query doesn't provide a good way to type it and types it as `unknown` under the hood. When I try to type `payload`, I keep on getting variants of `unknown` is not assignable to `User.AsObject` ...

lucaslcode replied:
I see, how about moving the type assertion into the hook instead? Only if it's easy.

In an closed stale issue they mention they might accept a PR they puts generic types in, I might try that. Could be straightforward.

darrenvong replied:
Aha it's a bit of a hack, as if they ever update the signature, we'd have to update the types ourselves as well, but I managed to augment/extend the `react-query` types to allow `useQueries` to be typed with generics, so no more `as` casting/type assertions!

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.ts:
```diff
@@ -0,0 +1,47 @@
+      );
+    } else {
+      errors = error && error.message ? [error.message] : [];
+    }
```
Can this be simplified to something like

```
const errors = [error?.message , ... friendQueries.map((q) => q.error?.message).where((e) => e !== null);
```

lucaslcode replied:
Across multiple lines ofc, I'm on my phone

lucaslcode reviewed (commented):
Think this looks really nice, let's discuss more in the meeting

darrenvong committed:
feat: simplify error building code across multiple queries

darrenvong committed:
feat: Move user type conversion logic into the useFriendList hook

darrenvong committed:
feat: augment react-query module

To allow useQueries to be typed with generics. Updated consuming code
with example usage

darrenvong commented:
Closing since this will be replaced by the React Query implementation in the demo PR as agreed in the last meeting.

darrenvong committed:
feat: add useFriendList tests

Removed the manual mock as it made debugging harder and it doesn't do
much anyway since the mocks get reset (including its implementation)
after the first test in the test suite anyway. And it's better to keep
the reset behaviour to ensure the mocks don't leak between tests.

darrenvong commented:
Finally got around to wrapping this up with some tests, ready to be reviewed again

darrenvong commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -0,0 +1,202 @@
+const getUserMock = service.user.getUser as jest.Mock;
+
+const wrapper = ({ children }: { children: React.ReactNode }) => {
+  const client = new QueryClient({
```
Inlining this in the wrapper is key to make sure a new client/cache is created for each test so they are isolated. This was causing me so much headaches before I realised!

darrenvong commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -1,101 +0,0 @@
-import { service as originalService } from "..";
```
Deleted this as this was also causing a lot of headache when trying to debug, since console logging a proxy gives you back a Symbol object...

Jest mocking correctly mocks out everything deeply anyway when we do `jest.mock` on this, so this is unnecessary code and it's interfering with the good default. And since the implementations in each mocks get reset, this is only good for one test anyway so it's useless. I've moved the default implementations into its own file in the `test` folder where all the test helpers are instead.

lucaslcode replied:
Sure. How are most of the tests still passing without changes? Because they didn't really use this?

darrenvong replied:
Yeah this wasn't being used, the existing tests either set the mock implementation before each test or had to manually override the default mock implementation to trigger certain scenarios (e.g. throwing an error).

lucaslcode commented on app/frontend/src/App.css:
```diff
@@ -1 +1,5 @@
 @import url("https://fonts.googleapis.com/css2?family=Lato:wght@400;700&display=swap");
+
+* {
+  box-sizing: border-box;
```
Why this? Not sure we should start using this file.

darrenvong replied:
This was actually for Storybook as the layout was using `content-box` for some strange reasons. I'll look at moving this into a component on the global Storybook decorator instead.

lucaslcode commented on app/frontend/src/features/connections/friends/FriendList.tsx:
```diff
@@ -0,0 +1,89 @@
+          Your friends
+        </Typography>
+        {isLoading ? (
+          <CircularProgress className={baseClasses.circularProgress} />
```
Is this style specific to the connections `CircularProgress`?

darrenvong replied:
Yeah, it's specifically for centering it to the middle of the tile/container

lucaslcode commented on app/frontend/src/features/connections/ConnectionsPage.tsx:
```diff
@@ -0,0 +1,28 @@
+  return (
+    <>
+      <PageTitle>My Connections</PageTitle>
+      <TabContext value={value}>
+        <TabBar value={value} setValue={setValue} labels={labels} />
+        <TabPanel value="TAB_ALL">ALL</TabPanel>
+        <TabPanel value="TAB_FRIENDS">
+          <FriendsTab />
+        </TabPanel>
+      </TabContext>
```
Is there an easy way to make tabs "routes"? Maybe a `<Switch>` instead of the tab panels? Would probably need a `useEffect` to set `value` too. This is also the case for the messages page, later.

darrenvong replied:
This was copied/adapted from the messages stuff which looks not too different from the example code in the Material UI page, so not sure without playing around with it a bit more tbh.

I did think that was annoying for messages though as currently there's no way to copy/paste a link that would get you straight to the group chat, but that might be for a separate PR/issue.

darrenvong replied:
Addressed this for the friend list tab by using URL params to manage state. The cool thing about that is that if the user visits a specific URL, it will jump straight to that tab rather than always defaulting to the first tab.

I think I'll leave the message tabs as a future fix.

lucaslcode commented on app/frontend/src/features/connections/friends/FriendList.stories.tsx:
```diff
@@ -0,0 +1,29 @@
+  component: FriendList,
+  decorators: [
+    (Story) => (
+      <Provider store={store}>
```
Is there a store needed in here?

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -0,0 +1,202 @@
+    });
+    expect(getUserMock).not.toHaveBeenCalled();
+
+    await waitForNextUpdate();
```
Why is waiting necessary after the assertions?

darrenvong replied:
This was there to get rid of that warning which says there's state updating outside of `act`, even though we don't care it about it in this case since we are interested in the loading state in this test rather than the eventual state when things get resolved successfully.

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -0,0 +1,202 @@
+    });
+    await waitForNextUpdate();
+
+    expect(getUserMock).toHaveBeenCalledTimes(2);
```
Is this because the default test user has 2 friends? If so maybe put a comment.

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -0,0 +1,202 @@
+      ],
+    });
+  });
+});
```
I am going to steal these for `useUsers` tests :D

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.ts:
```diff
@@ -0,0 +1,40 @@
+    isLoading: isLoading || friendQueries.some((query) => query.isLoading),
+    isError: hasErrors,
+    errors,
+    friendQueries,
```
I think we should be returning the data rather than the queries. The info on the queries is covered by isLoading, isError and errors. Then it is more similar to `useQuery`

darrenvong replied:
Good point. Although that got me thinking, and I haven't done it at the moment, but do we want to ever display errors at a granular level? For example, some friends loaded successfully but 1-2 errored for some reason. 

Or do we only display their friends when everyone has successfully loaded, otherwise we display an error over the panel/tile?

lucaslcode replied:
I think in this case it's not necessary.

Hmm... Unless the error is there's no connection. What happens in that case and there are still cached users? `isError` is true but data is defined? That is something I need to think about for `useUsers` too.

lucaslcode replied:
If the returned data is a `Map<number, User.AsObject|undefoned>` we could display unloaded users as "(error loading user)" as well as the error message. 

Anyway do what you think is best/easiest for now. We will probably change this to use `useUsers` once that's merged and we can think about it then.

darrenvong replied:
Yeah `isError` will be true with data defined in case of cached user, which is a bit of a weird state to be in. I imagine in that case we'd want to keep displaying the user?

Only brought this up because of how I defined `isError` here too. I went with the optimistic definition of only saying there is an error if everything errored, but that definitely complicate things.

A blanket error might be the easiest approach for now (although might not be best UX?) and like you said we can re-visit later.

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -0,0 +1,202 @@
+    });
+    await waitForNextUpdate();
+
+    expect(result.current).toMatchObject({
```
This isn't passing

lucaslcode reviewed (changes requested):
Nice work, only some relatively minor stuff

darrenvong committed:
feat: fix failing test

darrenvong committed:
feat: add missing QueryClientProvider to story file

Also moved Storybook specific CSS required into the .storybook setup
folder

darrenvong committed:
feat: merge conflict fix

Moved AuthProvider as child of Router since it uses `useHistory`.

Moved new mock listHostRequests implementation into the new
`serviceMockDefaults` helper file

darrenvong committed:
feat: change useFriendList to return closer to react-query's API

darrenvong committed:
feat: lift state to URL params

This has the benefit of allowing the user to jump straight to a tab
with a specific URL.

darrenvong committed:
feat: derive isError from length of errors list

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -86,39 +84,31 @@ describe("when the listFriends query succeeds", () => {
       wrapper,
     });
-    await waitForNextUpdate();
+    await waitForNextUpdate({ timeout: 10 });
```
Was this necessary to not fail? I'd rather avoid waiting in tests. Not sure if that's possible though.

darrenvong replied:
Yeah that wait can't be avoided, since any promise based ops (even ones that have already resolved) won't get its result until the next tick of the event loop, so the wait lets this run for one tick. It's not really a problem if the promise will resolve straight away - most of these tests take like ms to run.

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -86,39 +84,31 @@ describe("when the listFriends query succeeds", () => {
       return userId === "2"
         ? getUser(userId)
-        : new Promise((r) => setTimeout(() => r(getUser(userId)), 1));
+        : new Promise((r) => setTimeout(() => r(getUser(userId)), 50));
```
```suggestion
        : new Promise(() => null);
```

Would that make a never-resolving promise, or would it just throw?

darrenvong replied:
I thought this might cause some weird timeout from jest but it actually doesn't, think I'll change it to this when I get some time later so I don't need to specify a time for `waitForNextUpdate` to wait. It seems to know to wait for only resolved promises and not pending ones so will make the test a bit clearer than randomly waiting for 10ms as a hack.

lucaslcode commented on app/frontend/src/features/connections/friends/useFriendList.test.tsx:
```diff
@@ -195,8 +166,61 @@ describe("when the listFriends query failed", () => {
+    expect(result.current).toMatchObject({
+      isLoading: false,
+      isError: true,
+      errors: [
+        "Error listing friends",
+        "Error fetching user data",
+        "Error fetching user data",
+      ],
```
Hmm this means if the user's connection drops and they have 20 friends and they try to load the friends list, they will get an alert box with 21 lines. What do you think? Should we just have one "error" per hook and print individual errors to the console with `console.error()`?

lucaslcode replied:
Could save for another pr if you want.

darrenvong replied:
Yeah, I think I'll leave this for when we might refactor this hook to use your `useUsers` hook

lucaslcode reviewed (dismissed):
lgtm, feel free to ignore comments

darrenvong committed:
feat: change test for a user to return forever pending promise

So waitForNextUpdate doesn't need a specific time to wait for.
It'll know to wait for one tick only.

darrenvong commented:
Sorry, made the small tweaks you suggested so it dismissed the approval 😅
